Rewrite

benkelaar opened the pull request:
Turned it into pull request for comments etc.

Main thing missing is a lot of variable declarations. Both on global variables but especially on objects that should be local but are now initialized as fields on the global object.

The clearer split of concerns and packages is a lot better though, is nice

Treora committed:
Start rebuilding from scratch in /cothink2

Treora committed:
Update meteor

Treora committed:
Add assert, pass around Item objects

Treora committed:
Set state directly

To prevent 'undefined' errors.

Treora committed:
Create first autorun for updating layout

Visiting /item/123 triggers 'render' of item 123.

Treora committed:
Some renaming

Treora committed:
Create visibleItems collection.

The collection does not contain the item content, but only their id,
because the collection cannot store reactive objects. getVisibleItems
maps each id in the collection to the actual item, and is thus reactive
both to changes in the visibleItems collection and to the items
themselves.

Treora committed:
Add Mongo collection to store items serverside

Treora committed:
Move itemManager into own file

Treora committed:
Reintroduce CSS

Treora committed:
Set focussed class on focussed item

Currently checking uiCore.state.equals('focussedItem', this._id);
Should an item read uiCore state directly? uiCore.state should have a
getFocussedItem getter. Also maybe item should have a isFocussed getter
in the Item instance (in its state?) rather than in the template helper.

Treora committed:
Remove template helpers for isFocussed

Let's not use template helpers for things that should be object
attributes.

Treora committed:
Keep item data in Item.getData

Abandon the pattern introduced earlier; a function like getVisibleItems
should not be reactive to both the selection of items and the items'
contents, because changing an item's content should only trigger an
update for that single item. Getting the content should be done in a
separate autorun (by the template/Blaze).

Treora committed:
Show raw content

Treora committed:
Make items draggable

Treora committed:
Set item's _id attribute

Needed for {{#each}} to not mix up which item is which.

benkelaar commented on cothink2/client/itemmanager.js:
```diff
@@ -0,0 +1,14 @@
+itemManager = (function () {
+
+    var getItemById = function (id) {
```
When you use this construct to provide an unexposed object-wide scope, you could use named functions, preference thing I guess.

benkelaar commented on cothink2/client/item.js:
```diff
@@ -0,0 +1,16 @@
+Item = {};
```
Missing variable initialization (or does Meteor fix that for us?)

Treora replied:
Meteor fixes that. Variables without declaration become global (app-wide) variables. Declared variables are scoped to their file.

benkelaar commented on cothink2/client/itemmanager.js:
```diff
@@ -0,0 +1,14 @@
+itemManager = (function () {
```
var

benkelaar commented:
Huh, if Meteor wraps your javascript then perhaps the auto-initialized fields on the global object are the only way to share data, is that what's going on?

benkelaar commented on cothink2/client/lib/helpers.js:
```diff
@@ -0,0 +1,11 @@
+assert_type = function (object, prototype) {
+    message = "Object " + object + " is not of type " + prototype + ".";
+    assert(prototype.isPrototypeOf(object), message);
+}
```
Maybe use http://docs.meteor.com/#/full/check and/or http://docs.meteor.com/#/full/match_test

Treora replied:
Thanks for the tip, but simply `check(item, Item)` would not work: we would need to define something like `ItemFields = {id: String, _id: String}` to use for checking. It would allow duck typing but it would not check for the presence of methods, so it is probably not the right tool here.

I'd be happy to switch to other approaches, I think some type checking can be helpful to detect mistakes (something like Java's interfaces may be nicer than a class-like system though, which my `assert_type` solution tends towards).

benkelaar commented on cothink2/client/messagebus.js:
```diff
@@ -0,0 +1,44 @@
+ */
+
+msgbus = function (jQuery) {
+    channels = {}
```
Here you're definitely missing the variable declarations right? Also, why does this object initialization need jQuery?

Treora committed:
Move rewrite from cothink2 to cothink

Why did I make it a separate directory in the first place, we invented
version control for a reason. :)
